File: UCFPT/highschoolonlinedec2021/contest/solutions/hangman/hangman.java
package solutions.hangman;

import java.util.*;

public class hangman{
    public static boolean isValid(String q, String w){
        
        //If the query and the word are not the same length, it cannot be this word
        if(q.length()!=w.length()) return false;
        
        //Initialize boolean arrays to check if a letter has been called or not
        boolean letterUsed[] = new boolean[26];
        boolean isOpen[] = new boolean[26];

        //loop through word.
        for(int i = 0; i < q.length(); i++){

            //if we have a letter, check if it is the same and mark it as used, else mark it as open
            if(q.charAt(i) != '_'){
                if(q.charAt(i)!=w.charAt(i)) return false;
                letterUsed[q.charAt(i)-'a'] = true;
            }
            else{
                isOpen[w.charAt(i)-'a'] = true;
            }
        }

        //Check if there is a contradiction (a letter that is both used and open)
        for(int i = 0; i < 26; i++)
            if(letterUsed[i] && isOpen[i])
                return false;

        //If nothing else has been returned, then the word must be valid
        return true;
    }

    public static void main(String[] args) {
        
        Scanner in = new Scanner(System.in);

        int c = in.nextInt();
        for(int cc = 0; cc < c; cc++){

            //get category
            String category = in.next();

            //get dictionary
            int n = in.nextInt();
            String dictionary[] = new String[n];
            for(int i = 0; i < n; i++)
                dictionary[i] = in.next();

            //get and process queries
            int g = in.nextInt();
            for(int i = 0; i < g; i++){

                //scan query
                String query = in.next();
                boolean good = false;

                //check if query works for at least one word in a dictionary
                for(String word : dictionary){
                    if(isValid(query, word)){
                        good = true;
                        break;
                    }
                }

                //Output the answer
                if(good)
                    System.out.println("No cheaters here!");
                else
                    System.out.println("That's not a word for a(n) " + category + "!");
            }
        }
        in.close();
    }
}

File: UCFPT/highschoolonlinedec2021/contest/solutions/hangman/hangman.cpp
#include <bits/stdc++.h>
using namespace std;

// try all strings that match the pattern with backtracking
// since there are at most 4 _ chars, then there are 26^4
// strings at most
bool query(int i, string& s, vector<bool>& legal, set<string>& dict) {
	if (i == s.length()) 
		return dict.count(s);

	if (s[i] != '_')
		return query(i + 1, s, legal, dict);

	for (int j = 0; j < 26; j++) 
		if (legal[j]) {
			s[i] = 'a' + j;

			if (query(i + 1, s, legal, dict))
				return true;
		}

	s[i] = '_';
	return false;
}

void solve() {   
	string category;
	cin >> category;

	int n;
	cin >> n;

	set<string> dict;

	for (int i = 0; i < n; i++) {
		string s;
		cin >> s;
		dict.insert(s);
	}

	int q;
	cin >> q;

	for (int i = 0; i < q; i++) {
		string s;
		cin >> s;

		vector<bool> legal(26, 1);
		for (char c : s)
			if (isalpha(c))
				legal[c - 'a'] = 0;

		if (query(0, s, legal, dict))
			cout << "No cheaters here!\n";
		else
			cout << "That's not a word for a(n) " << category << "!\n";
	}
} 

int main() {
	int t; cin >> t; 
	while (t--) solve();

	return 0;
}
